contracts-periphery/src/UmbraBatchSend.sol

```diff
@@ -24,69 +27,131 @@ contract UmbraBatchSend {
     bytes32 ciphertext;
   }
 
+  mapping(address => uint256) internal totalTransferAmountPerToken;
+  struct TransferSummary {
+    uint256 amount;
+    address tokenAddr;
+  }
+
   error ValueMismatch();
+  error TransferAmmountMismatch();
   event BatchSendExecuted(address indexed sender);
 
   constructor(IUmbra _umbra) {
     umbra = _umbra;
   }
 
   function batchSendEth(uint256 _tollCommitment, SendEth[] calldata _params) external payable {
-    if(msg.value != _sumValueFromEthSends(_tollCommitment, _params)) revert ValueMismatch();
+    if (msg.value != _sumValueFromEthSends(_tollCommitment, _params)) revert ValueMismatch();
     _batchSendEth(_tollCommitment, _params);
     emit BatchSendExecuted(msg.sender);
   }
 
-  function batchSendTokens(uint256 _tollCommitment, SendToken[] calldata _params) external payable {
-    if(msg.value != _tollCommitment * _params.length) revert ValueMismatch();
-    _batchSendTokens(_tollCommitment, _params);
+  function batchSendTokens(
+    uint256 _tollCommitment,
+    SendToken[] calldata _params,
+    TransferSummary[] calldata _transferSummary
+  ) public payable {
+    if (msg.value < _tollCommitment * _params.length) revert ValueMismatch();
+
+    for (uint256 i = 0; i < _params.length; i++) {
+      // Used later to validate total amounts are correctly provided
+      totalTransferAmountPerToken[_params[i].tokenAddr] += _params[i].amount;
+    }
+
+    uint256 _summaryLength = _transferSummary.length;
+    for (uint256 i = 0; i < _summaryLength; ) {
+      IERC20 _token = IERC20(address(_transferSummary[i].tokenAddr));
+
+      if (totalTransferAmountPerToken[_transferSummary[i].tokenAddr] != _transferSummary[i].amount) {
+        revert TransferAmmountMismatch();
+      }
+
+      SafeERC20.safeTransferFrom(_token, msg.sender, address(this), _transferSummary[i].amount);
+      unchecked {
+        i++;
+      }
+    }
+    _batchSendTokens(_tollCommitment, _params, _transferSummary);
+
     emit BatchSendExecuted(msg.sender);
   }
 
   function batchSend(
     uint256 _tollCommitment,
     SendEth[] calldata _ethParams,
-    SendToken[] calldata _tokenParams
+    SendToken[] calldata _tokenParams,
+    TransferSummary[] calldata _transferSummary
   ) external payable {
     uint256 _totalAmount = _sumValueFromEthSends(_tollCommitment, _ethParams) + (_tollCommitment * _tokenParams.length);
-    if(msg.value != _totalAmount) revert ValueMismatch();
+    if (msg.value != _totalAmount) revert ValueMismatch();
 
     _batchSendEth(_tollCommitment, _ethParams);
-    _batchSendTokens(_tollCommitment, _tokenParams);
+    batchSendTokens(_tollCommitment, _tokenParams, _transferSummary);
```

Comment:
Why do we call the internal batch send ETH method but the external batch send tokens method?

Reply:
But actually that might make less sense the more i think about it

Reply:
What if we extract the eth value checks into a method that takes two arrays: one of `SendEth[]`, one of `SendToken[]`. If we're only sending ETH, we pass an empty array of `SendToken` into the helper method. Then we can reuse the same validation logic and simplify things a bit

Reply:
Addressed here https://github.com/ScopeLift/umbra-protocol/pull/351/commits/11ee5e7c67649edb0601ab23fcc802dae11f56f8